### PR TITLE
Refactor and document dnsa pool.

### PR DESCRIPTION
**Pull request by AdamStepan**

### CODE REVIEW

File: include/superior_mysqlpp/dnsa_connection_pool.hpp [hunk 1]
```diff
@@ -193,6 +228,9 @@ namespace SuperiorMySqlpp
                 enabled = false;
```
Comment (MrWarlockX) [inline include/superior_mysqlpp/dnsa_connection_pool.hpp:244]: This seems to me like duplicit operation (if condition *jobThread.joinable()* passes)

Comment (AdamStepan) [reply include/superior_mysqlpp/dnsa_connection_pool.hpp:244]: You are right, it should be in "else" branch.

File: include/superior_mysqlpp/dnsa_connection_pool.hpp [hunk 2]
```diff
@@ -201,21 +239,34 @@ namespace SuperiorMySqlpp
                 jobThread = std::thread{&DnsAwarePoolManagement::job, std::cref(*this)};
             }
 
+            /**
+             * Returns true if jobThread is in joinable state.
+             */
             auto isDnsAwarePoolManagementThreadRunning() const
             {
+                // XXX: this is ambigious, it can return false if jobThread is detached
```
Comment (OpatrilPeter) [inline include/superior_mysqlpp/dnsa_connection_pool.hpp]: Your point is correct, however  `jobThread` is private and I can't see any detaching - so it's more like a note about possible danger? I'd reformulate it that way.

File: include/superior_mysqlpp/dnsa_connection_pool.hpp [hunk 3]
```diff
@@ -201,21 +239,34 @@ namespace SuperiorMySqlpp
                 jobThread = std::thread{&DnsAwarePoolManagement::job, std::cref(*this)};
             }
 
+            /**
+             * Returns true if jobThread is in joinable state.
```
Comment (OpatrilPeter) [inline include/superior_mysqlpp/dnsa_connection_pool.hpp]: IMHO point of this function is checking if management thread is running - how it does so is implementation detail. See comment below on discussion about implementation.

File: include/superior_mysqlpp/dnsa_connection_pool.hpp [hunk 4]
```diff
@@ -63,15 +72,23 @@ namespace SuperiorMySqlpp
             }
         };
 
-
+        /**
+         * Mix in class that checks and logs DNS changes.
```
Comment (OpatrilPeter) [inline include/superior_mysqlpp/dnsa_connection_pool.hpp]: Mixin

File: include/superior_mysqlpp/dnsa_connection_pool.hpp (hunk 2 above)
Comment (OpatrilPeter) [reply include/superior_mysqlpp/dnsa_connection_pool.hpp]: `ambiguous`

File: include/superior_mysqlpp/dnsa_connection_pool.hpp (hunk 4 above)
Comment (AdamStepan) [reply include/superior_mysqlpp/dnsa_connection_pool.hpp]: ACK

File: include/superior_mysqlpp/dnsa_connection_pool.hpp [hunk 5]
```diff
@@ -101,88 +125,99 @@ namespace SuperiorMySqlpp
...
             {
                 return *static_cast<const Base*>(this);
             }
 
-            /*
+            /**
+             * This function is called when unexpected error is thrown during dns resolution error handling.
+             * It's called only if logger throws exception.
+             */
+            void onUnexpectedError(void) const noexcept {
```
Comment (OpatrilPeter) [inline include/superior_mysqlpp/dnsa_connection_pool.hpp]: Please use bracketing consistent with the rest of the code (that includes `catch` etc.)

Comment (AdamStepan) [reply include/superior_mysqlpp/dnsa_connection_pool.hpp]: ACK

File: include/superior_mysqlpp/dnsa_connection_pool.hpp (hunk 2 above)
Comment (AdamStepan) [reply include/superior_mysqlpp/dnsa_connection_pool.hpp]: ACK

File: include/superior_mysqlpp/dnsa_connection_pool.hpp [hunk 6]
```diff
@@ -101,88 +125,99 @@ namespace SuperiorMySqlpp
...
+                }
+            }
+
+            /**
              * This function is run as parallel thread.
              * We cannot afford to let any exceptions to be propagated from logging functions
-             * since it would result in call to std::terminate()!!! Hence all the try/catch blocks.
+             * since it would result in call to std::terminate().
+             */
+            void job(void) const {
```
Comment (OpatrilPeter) [inline include/superior_mysqlpp/dnsa_connection_pool.hpp]: Explicit void in parameters is completely redundant in C++ (not in C)

Comment (AdamStepan) [reply include/superior_mysqlpp/dnsa_connection_pool.hpp]: ACK

File: include/superior_mysqlpp/dnsa_connection_pool.hpp [hunk 7]
```diff
@@ -39,6 +41,13 @@ namespace SuperiorMySqlpp
             DnsResolver& operator=(const DnsResolver&) = default;
             DnsResolver& operator=(DnsResolver&&) = default;
 
+            /**
+             * Resolve a hostname to ip addresses.
+             *
+             * @tparam const reference to std::string
+             * @param hostname
+             * @return Vector of strings that represent ip addresses for given hostname.
+             */
```
Comment (majcl) [inline include/superior_mysqlpp/dnsa_connection_pool.hpp]: I'm wondering if it wouldn't be nice whether methods docs should have throws section. 
At least here, IMHO it makes sense to mention possible asio exceptions.

Comment (majcl) [reply include/superior_mysqlpp/dnsa_connection_pool.hpp]: What do you think?
Also @OpatrilPeter @smradCZ

File: include/superior_mysqlpp/dnsa_connection_pool.hpp [hunk 8]
```diff
@@ -81,6 +98,10 @@ namespace SuperiorMySqlpp
 
             DnsAwarePoolManagement(const DnsAwarePoolManagement&) = delete;
 
+            /**
+             * Initializes pool from other pool.
+             * It starts DNS resolution job if job is running in other instance.
```
Comment (majcl) [inline include/superior_mysqlpp/dnsa_connection_pool.hpp:109]: 👍

File: include/superior_mysqlpp/dnsa_connection_pool.hpp [hunk 9]
```diff
@@ -101,88 +125,106 @@ namespace SuperiorMySqlpp
...
             }
 
-            /*
+            /**
+             * This function is called when unexpected error is thrown during dns resolution error handling.
+             * It's called only if logger throws exception.
+             */
+            void onUnexpectedError() const noexcept
+            {
+                if(terminateOnFailure) {
```
Comment (majcl) [inline include/superior_mysqlpp/dnsa_connection_pool.hpp]: space after if

File: include/superior_mysqlpp/dnsa_connection_pool.hpp [hunk 10]
```diff
@@ -101,88 +125,106 @@ namespace SuperiorMySqlpp
...
-                        std::terminate();
-                    }
-                };
-
                 try
                 {
-                    std::vector<std::string> lastIpAddresses{};
+                    runJobLoop();
+                }
+                catch(std::exception &e)
```
Comment (majcl) [inline include/superior_mysqlpp/dnsa_connection_pool.hpp]: Also space after catches.

File: include/superior_mysqlpp/dnsa_connection_pool.hpp [hunk 11]
```diff
@@ -101,88 +125,106 @@ namespace SuperiorMySqlpp
...
+                if(terminateOnFailure) {
+                    std::terminate();
+                }
+            }
+
+            /**
              * This function is run as parallel thread.
              * We cannot afford to let any exceptions to be propagated from logging functions
-             * since it would result in call to std::terminate()!!! Hence all the try/catch blocks.
+             * since it would result in call to std::terminate().
```
Comment (majcl) [inline include/superior_mysqlpp/dnsa_connection_pool.hpp:164]: Right, but in outer try/catch block, we also try to log -> so the problem is still there, isn't it?

File: include/superior_mysqlpp/dnsa_connection_pool.hpp [hunk 12]
```diff
@@ -101,88 +125,106 @@ namespace SuperiorMySqlpp
...
-                    getBase().getLogger()->logSharedPtrPoolDnsAwarePoolManagementError(getBase().getId(), e);
-                    onError();
-                }
-                catch (...)
-                {
-                    getBase().getLogger()->logSharedPtrPoolDnsAwarePoolManagementError(getBase().getId());
-                    onError();
+                    getBase().getLogger()->logSharedPtrPoolDnsAwarePoolManagementCycleEnd(getBase().getId());
+
+                    sleepInParts(sleepTime, std::chrono::milliseconds{50}, [&](){ return enabled.load(); });
```
Comment (majcl) [inline include/superior_mysqlpp/dnsa_connection_pool.hpp]: I think that 50ms should probably be at least some kind of "explicit" constant?

**Review by majcl — CHANGES_REQUESTED** [review]

File: include/superior_mysqlpp/dnsa_connection_pool.hpp (hunk 10 above)
Comment (AdamStepan) [reply include/superior_mysqlpp/dnsa_connection_pool.hpp]: ACK

File: include/superior_mysqlpp/dnsa_connection_pool.hpp (hunk 9 above)
Comment (AdamStepan) [reply include/superior_mysqlpp/dnsa_connection_pool.hpp]: ACK

File: include/superior_mysqlpp/dnsa_connection_pool.hpp (hunk 7 above)
Comment (AdamStepan) [reply include/superior_mysqlpp/dnsa_connection_pool.hpp]: OK, I'll to document all exceptions which are thrown by asio.

File: include/superior_mysqlpp/dnsa_connection_pool.hpp (hunk 11 above)
Comment (AdamStepan) [reply include/superior_mysqlpp/dnsa_connection_pool.hpp:164]: Yes, the problem is still there. I will remove logger's logging from the outer catch block.

File: include/superior_mysqlpp/dnsa_connection_pool.hpp (hunk 12 above)
Comment (AdamStepan) [reply include/superior_mysqlpp/dnsa_connection_pool.hpp]: What about "shutdownCheckPeriod"?

**Pull request closed without merging**